src/main/java/io/cdap/plugin/http/source/common/RawStringPerLine.java

```diff
@@ -61,14 +65,26 @@ public boolean hasNext() {
             isLineRead = true;
             return lastLine != null;
         } catch (IOException e) { // we need to catch this, since hasNext() does not have "throws" in parent
-            throw new RuntimeException("Failed to read line from http page buffer", e);
+            ErrorUtils.ActionErrorPair pair = ErrorUtils.getActionErrorByStatusCode(httpResponse.getStatusCode());
+            String errorMessage = String.format("Unable to read line from http page buffer: %s. %s",
+              httpResponse.getStatusCode(), pair.getCorrectiveAction());
+            throw ErrorUtils.getProgramFailureException(new ErrorCategory(ErrorCategory.ErrorCategoryEnum.PLUGIN),
+                    errorMessage, errorMessage, pair.getErrorType(), true, ErrorCodeType.HTTP,
+              String.valueOf(httpResponse.getStatusCode()),
+              HttpErrorDetailsProvider.getSupportedDocumentUrl(), e);
         }
     }
 
     @Override
     public String next() {
         if (!hasNext()) { // calling hasNext will also read the line;
-            throw new NoSuchElementException();
+            ErrorUtils.ActionErrorPair pair = ErrorUtils.getActionErrorByStatusCode(httpResponse.getStatusCode());
+            String errorMessage = String.format("Failed to read the next line with error code: %s. %s",
+                    httpResponse.getStatusCode(), pair.getCorrectiveAction());
+            throw ErrorUtils.getProgramFailureException(new ErrorCategory(ErrorCategory.ErrorCategoryEnum.PLUGIN),
+                    errorMessage, errorMessage, pair.getErrorType(), true, ErrorCodeType.HTTP,
+                    String.valueOf(httpResponse.getStatusCode()),
+                    HttpErrorDetailsProvider.getSupportedDocumentUrl(), null);
```

Comment:
It is not necessary `hasNext()` is `false`, the request failed, it might not have any element to read.

please do a check if `httpResponseCode` is not `200` then only use `httpResponseCode`.

Otherwise throw a generic ProgramFailureException.